### PR TITLE
Hydra Model updated

### PR DESCRIPTION
**Pull request by tomcrane**
This updates the develop branch with changes to the Hydra model that will be used by the updated API and Portal.

I have started a DLCS.Hydra.Tests project but it only contains some simple tests for now.

The DLCS.Mock project (adapted from the 2015 original) is a kind of test of this model, but it's not a runnable test like the others. It's not unit testing!

So we need to think how the Hydra base library and our DLCS.HydraModel are tested (I don't think we need separate test projects for these).

While there needs to be more Hydra testing I felt like I was making fairly trivial tests at the moment. 
I have removed some, but they can come back when testing API responses.

### CODE REVIEW

File: API.Client/HydraHttpResponseMessageX.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using DLCS.Web.Response;
+using Hydra;
+using Hydra.Model;
+using Newtonsoft.Json;
+
+namespace API.Client
+{
+    public static class HydraHttpResponseMessageX
+    {
+        public static async Task<T?> ReadAsHydraResponseAsync<T>(this HttpResponseMessage response,
```
Comment (donaldgray) [inline API.Client/HydraHttpResponseMessageX.cs:14]: xml-comment here?

File: DLCS.Hydra.Tests/Model/ImageTests.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        var expected = BaseUrl + ImageApiPath;
+        var image = new Image(BaseUrl, 1, 1, "my-image");
+        var jsonString = JsonConvert.SerializeObject(image, serializerSettings);
+        var jObject = JObject.Parse(jsonString);
+        jObject["@id"].Value<string>().Should().Be(expected);
+    }
+    
+    [Fact]
+    public void ImageClass_SerialisesTo_shortStringForImage_Id()
```
Comment (donaldgray) [inline DLCS.Hydra.Tests/Model/ImageTests.cs:44]: Small thing `ImageClass_SerialisesTo_shortStringForImage_Id` -> `ImageClass_SerialisesTo_ShortStringForImage_Id`

File: DLCS.HydraModel/Customer.cs [hunk 3]
```diff
@@ -135,9 +168,18 @@ public override void DefineOperations()
...
                 .GetStandardCollectionOperations(operationId + "space_", "Space", "vocab:Space");
 
+            
+            var images = GetHydraLinkProperty("allImages");
+            images.SupportedOperations = CommonOperations
+                .GetStandardCollectionOperations("_:customer_image_", "Image", "vocab:Image");
+            images.SupportedOperations.WithMethod("GET").Description =
+                "Can take query parameters";
+            images.SupportedOperations.WithMethod("POST").Description =
+                "Push an image for immediate processing, asynchronously. Might fail or timeout. This operation is rate-limited.";
```
Comment (donaldgray) [inline DLCS.HydraModel/Customer.cs:181]: `"This operation is rate-limited."` is it? Or should it be?

File: DLCS.HydraModel/DlcsResource.cs [hunk 4]
```diff
@@ -68,8 +68,8 @@ public virtual JObject GetCollectionForm()
 
         public override string Context =>
             // do this better later - prefer not have FullName but makes reflection easier!
-            BaseUrl + "/contexts/" + GetType().FullName + ".jsonld";
+            BaseUrl + "/contexts/" + GetType().Name + ".jsonld";
```
Comment (donaldgray) [inline DLCS.HydraModel/DlcsResource.cs:71]: Use string interpolation here for consistency?

File: Hydra/HydraStringX.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+        public static int? GetLastPathElementAsInt(this string? path)
+        {
+            var last = path.GetLastPathElement();
+            if (string.IsNullOrWhiteSpace(last))
+            {
+                return null;
+            }
+
+            // We want this to throw if not an int
```
Comment (donaldgray) [inline Hydra/HydraStringX.cs:35]: Add a test for this, to confirm behaviour? Something like

```cs
var path = "foo/bar";
Action action = () => path.GetLastPathElementAsInt();
action.Should().Throw<InvalidCastException>()
```

**Review by donaldgray — APPROVED** [review]
Couple of comments but all minor

File: DLCS.HydraModel/DlcsResource.cs (hunk 4 above)
Comment (tomcrane) [reply DLCS.HydraModel/DlcsResource.cs:71]: Yeah a lot of Hydra code here from 2016, will fix up

**Pull request merged**